core_backend/app/llm_call/llm_prompts.py

```diff
@@ -292,29 +292,100 @@ class AlignmentScore(BaseModel):
     ).strip()
 
 
-def get_urgency_detection_prompt(condition: str, message: str) -> str:
+class UrgencyDetectionEntailment:
     """
-    Returns the prompt for the urgency detection bot.
+    Urgency detection using entailment.
     """
 
-    return textwrap.dedent(
-        (
-            """Given a [statement] and [comment], score the share of meaning
-            of [statement] covered by [comment].
-            Respond with a score between 0 and 1 with 0.1 increments.
-            """
-            """
-            Respond in json string:
+    class UrgencyDetectionEntailmentResult(BaseModel):
+        """
+        Pydantic model for the output of the urgency detection entailment task.
+        """
 
-            \{
-               statement: str
-               probability: float
-               reason: str
-            \}
-            """
-            f"""
-            statement: {condition}
-            comment: {message}
-            """
-        )
+        best_matching_rule: str
+        probability: float = Field(ge=0, le=1)
+        reason: str
+
+    _urgency_rules: List[str]
+    _prompt_base: str = textwrap.dedent(
+        """
+        You are a highly sensitive urgency detector. Score if ANY part of the
+        user message corresponds to any part of the urgency rules provided below.
+        Ignore any part of the user message that does not correspond to the rules.
+        Respond with (a) the rule that is most consistent with the user message,
+        (b) the probability between 0 and 1 with increments of 0.1 that ANY part of
+        the user message matches the rule, and (c) the reason for the probability.
+
+
+        Respond in json string:
+
+        {
+           best_matching_rule: str
+           probability: float
+           reason: str
+        }
+        """
+    ).strip()
+
+    _prompt_rules: str = textwrap.dedent(
+        """
+        Urgency Rules:
+        {urgency_rules}
+        """
     ).strip()
+
+    default_json: Dict = {
+        "best_matching_rule": "",
+        "probability": 0.0,
+        "reason": "",
+    }
+
+    def __init__(self, urgency_rules: List[str]) -> None:
+        """
+        Initialize the urgency detection entailment task with urgency rules.
+        """
+        self._urgency_rules = urgency_rules
+
+    def parse_json(self, json_str: str) -> Dict:
+        """
+        Validates the output of the urgency detection entailment task.
+        """
+
+        json_str = json_str.replace("```json", "").replace("```", "")
+        json_str = json_str.replace("\{", "{").replace("\}", "}")
+
+        # fmt: off
+        json = (
+            UrgencyDetectionEntailment
+                .UrgencyDetectionEntailmentResult
+                .model_validate_json(
+                    json_str
+                )
+            )
+        # fmt: on
+
+        if json.best_matching_rule not in self._urgency_rules:
```

Comment:
<img width="916" alt="image" src="https://github.com/IDinsight/aaq-core/assets/7042047/b8e08336-f42e-4222-8b7d-a0a41608dc83">
The rule was exactly "How many advils" -- should we add a clean up step for this field?

Also, when the error is raised, we return the default response. We should have an error response for urgency detection. And/or at least update the response json with valid json fields maybe as a temporary solution?

```json
{
  "is_urgent": false,
  "failed_rules": [],
  "details": {
    "best_matching_rule": "",
    "probability": 0,
    "reason": ""
  }
}
```

Reply:
Added a regex to strip the number from it. Thanks for picking this up.

And when it fails, it raises a warning and returns default json - see `detect_urgency` in `core_backend/app/llm_call/entailment.py`. That's actually what is throwing the JSONDecode Error.